src/rimeengine.cpp

```diff
@@ -323,14 +416,118 @@ void RimeEngine::updateConfig() {
+    }
+    while (api_->config_next(&iter)) {
+        RimeConfigIterator subIter;
+        std::string path = iter.path;
+        std::string statesPath = path + "/states";
+        if (!api_->config_begin_list(&subIter, &config, statesPath.c_str())) {
+            continue;
+        }
+        std::vector<std::string> labels;
+        const char *value = nullptr;
+        while (api_->config_next(&subIter)) {
+            value = api_->config_get_cstring(&config, subIter.path);
+            if (!value) {
+                break;
+            }
+            labels.emplace_back(value);
+        }
+        if (!value || labels.size() <= 1) {
+            continue;
+        }
+        std::string namePath = path + "/name";
+        const char *name = api_->config_get_cstring(&config, namePath.c_str());
+        if (name) {
+            if (labels.size() != 2) {
+                continue;
+            }
+            std::string optionName = name;
+            if (optionName == RIME_ASCII_MODE) {
+                // imAction_ has latin mode that does the same
+                continue;
+            }
+            optionActions_.emplace_back(std::make_unique<ToggleAction>(
+                this, optionName, labels[0], labels[1]));
+            instance_->userInterfaceManager().registerAction(
+                "fcitx-rime-" + currentSchema + "-" + optionName,
+                optionActions_.back().get());
+        } else {
+            std::string optionsPath = path + "/options";
+            if (!api_->config_begin_list(&subIter, &config,
+                                         optionsPath.c_str())) {
+                continue;
+            }
+            std::vector<std::string> options;
+            value = nullptr;
+            while (api_->config_next(&subIter)) {
+                value = api_->config_get_cstring(&config, subIter.path);
+                if (!value) {
+                    break;
+                }
+                options.emplace_back(value);
+            }
+            if (!value || labels.size() <= 1 ||
+                labels.size() != options.size()) {
+                continue;
+            }
+            optionActions_.emplace_back(std::make_unique<SelectAction>(
+                this, currentSchema, options, labels));
+            instance_->userInterfaceManager().registerAction(
+                "fcitx-rime-" + currentSchema + "-select-" + options[0],
+                optionActions_.back().get());
+        }
+        statusArea.addAction(StatusGroup::InputMethod,
+                             optionActions_.back().get());
+    }
+
+fail:
+    api_->config_close(&config);
+}
+
+void RimeEngine::refreshStatusArea(RimeSessionId session) {
+    instance_->inputContextManager().foreachFocused(
+        [this, session](InputContext *ic) {
+            if (auto state = this->state(ic)) {
+                // After a deployment, param is 0, refresh all
+                if (!session || state->session() == session) {
```

Comment:
state->session() would request a new session if there's no existing session. You may want to add a bool to RimeState::session(bool requestNewSession = true), so you can get session without request new session.

Also, this function would also need to check if current input method is rime, otherwise it will mess up the action of other input method.